### ISSUE TITLE
"Invalid dst" when creating SSH ACL

### ISSUE THREAD

**Issue by mholt** [tailscale/tailscale#5084]
### What is the issue?

I'm trying to make it so that my family member can log into a machine on my network via SSH. I've shared the machine with them on Tailscale, and now want to add the ACL so they can login via SSH:

```json
		{
		  "action": "accept", // "accept" or "check"
		  "src": ["their@email"],
		  "dst": ["100.93.13.58"],
		  "users": ["root"]
		},
```

But every time I add this and click Save, I get:

> Error: invalid dst "100.93.13.58"

The IP is definitely one that is on my Machines list, and I've also tried setting its hostname in the "hosts" portion of the config and using that, but I get the same error even with a defined hostname.

### Tailscale version

v1.28

**Comment by mholt** [comment]
It does work if I use a tag instead. Working ACL:

```json
// Example/default ACLs for unrestricted connections.
{
	// Declare static groups of users beyond those in the identity service.
	"groups": {
		"group:example": ["user1@example.com", "user2@example.com"],
	},

	// Declare convenient hostname aliases to use in place of IP addresses.
	"hosts": {
		"raspberrypi": "100.93.13.58",
	},

	"tagOwners": {
		"tag:raspi": ["me@mine.com"],
	},

	// Access control lists.
	"acls": [
		// Match absolutely everything.
		// Comment this section out if you want to define specific restrictions.
		// {"action": "accept", "users": ["*"], "ports": ["*:*"]},
		{
			"action": "accept", // "accept" or "check"
			"src":    ["him@his.com"],
			"dst":    ["100.93.13.58:22"],
		},
	],
	"ssh": [
		// Allow all users to SSH into their own devices in check mode.
		// Comment this section out if you want to define specific restrictions.
		{
			"action": "check",
			"src":    ["autogroup:members"],
			"dst":    ["autogroup:self"],
			"users":  ["autogroup:nonroot", "root"],
		},
		{
			"action": "accept", // "accept" or "check"
			"src":    ["him@his.com"],
			"dst":    ["tag:raspi"],
			"users":  ["pi"],
		},
	],
}
```

However, simply changing `["tag:raspi"]` to `["100.93.13.58"]` or `["raspberrypi"]` causes the "invalid dst" error.

**Comment by mholt** [comment]
Oh... [this tweet](https://twitter.com/apenwarr/status/1549157729742331909) made me take a closer read of the docs:

> The only kinds of connections that are allowed are:
> 
>  - From a user to their own devices, as any user including root.
>  - From a user to a tagged device, as any user including root.
>  - From a tagged device to another tagged device, for any tags. Note that an SSH access rule from a tagged device cannot be in [check mode](https://tailscale.com/kb/1193/tailscale-ssh/#configure-tailscale-ssh-with-check-mode).

Maybe the error is because SSH destinations can only be tagged devices? (Although, the first point says "from a user to their own devices" which leads me to believe that if "src" is myself, then I should be able to put a host or IP there instead of only being able to use a tag.)

So maybe I'm just a bit confused :sweat_smile: -- it feels like a bug though. And even if it isn't, maybe a more descriptive/helpful error message could be good? Thanks!

**Comment by mayakacz** [comment]
From https://tailscale.com/kb/1018/acls/#tailscale-ssh:
> The `dst` field is the destination where the connection goes. This can be a user, group, or tag.

Agree error could be improved.

I'm guessing you're trying to write a rule to access some of your devices as `dst` but not others? Can you explain why?

**Comment by mholt** [comment]
@mayakacz Ah, yeah, I did eventually find that. I was hoping to grant another Tailscale user access to one of my devices via SSH by sharing the device and setting the `dst` to the device's IP address.

I eventually figured out that I have to use a tag (which is less convenient, since I want to just use the IP address) and that I can't give access to another user. So, perhaps my quest is in vain.

So yeah, at this point:

- More descriptive error message would be nice
- Feature request: Share devices to grant SSH access to other users
- Feature request: Specify IP address in `dst`

Thank you :+1:

**Comment by maisem** [comment]
So having IPs in SSH is ACLs is deliberately restricted.
The restriction we are working around is that Alice should never be able to login to machine owned by Bob.

Say you have a machine M1 that is currently tagged, and then we let you add an IP based ACL for Alice. Sometime later that machine is untagged by Bob. This results in a policy where Alice can SSH into Bobs machine.

I agree that we should make the error and docs better.

**Comment by mholt** [comment]
Makes sense. In our case we're all family and we get along so trust is a bit higher and our threat model is a bit softer. Thanks!

**Comment by whitequark** [comment]
This is an incredibly annoying issue. Not being able to specify IPs means I'm much more likely to just use autogroup:self instead of trying to write a fine grained policy.
